### ISSUE TITLE
Leaves behind processes when restarting vscode on linux

### ISSUE THREAD

**Issue by alexriedl** [tintoy/msbuild-project-tools-vscode#36]
## Expected Behavior
Opening a C# project in vscode, and reloading the vscode window does not spawn additional extension processes

## Current Behavior
A new process is started each time vscode is reloaded and leaves old processes around

## Steps to Reproduce
1. Open a C# project in vscode 
2. View processes `ps -ef | grep msbuild-project`
3. In vscode, reload the window `ctrl+shift+p` -> `Reload Window`
4. View processes. Note there are now 2 instances of the extension running
5. Repeat step 3 to spawn additional processes

## Environment
kubuntu 17.10
msbuild project tools 0.2.27 (happened in older versions as well)

**Comment by alexriedl** [comment]
Its worth noting that this issue only exists with the default settings for the extension. If I change `useClassicProvider` to true, then no new processes are started when I restart vscode.

**Comment by tintoy** [comment]
Hi - sorry, I thought I'd fixed the last of these issues a while back but apparently not :-/

Does it also happen when completely exiting VS Code and then reopening it, or only when using the "reload window" command?

Would you mind setting a couple of options in configuration to generate a log file and reloading the window a couple of times?

```json
{
    "msbuildProjectTools": {
        "logging": {
            "level": "Verbose",
            "file": "/tmp/msbuild-project-tools.log",
        }
    }
}
```

If you can attach the resulting log file I'll see if I can work out what's going on (this doesn't seem to happen on Windows / MacOS but it's been a while since I tested on Linux).

**Comment by tintoy** [comment]
Actually, I just tried this again on my Mac and it _does_ leave a process behind. But not on Windows. Very odd. I have an idea of how to deal with this, but I may need to run an experiment or 2.

**Comment by tintoy** [comment]
It seems like the behaviour of VS Code when it comes to shutting down extensions / language servers is different on Windows vs Mac / Linux (although I'm still trying to narrow down the actual behaviour that differs).

I'll keep you posted.

**Comment by tintoy** [comment]
One other thing you might be able to try for me - on my Mac I got it to happen once or twice but not with the latest changes I made. But while debugging, I noticed that just before VS Code exits, if the MSBuild language service process is going to hang around, there's an unhandled exception in another extension (the `C#` extension).

Would you mind temporarily disabling the C# extension and then restarting VS Code, then seeing if you can still reproduce the problem with the MSBuild extension? This may help me eliminate one line of inquiry so it'd be pretty useful if you have the time.

**Comment by tintoy** [comment]
(it's enough to disable the `C#` extension, no need to uninstall it)

**Comment by alexriedl** [comment]
I currently have the latest released version of msbuild project tools installed (0.2.27). I confirmed the issue happens when reloading a vscode window as well as closing and reopening vscode altogether. After disabling the C# extension the issue still persists in both cases.

**Comment by alexriedl** [comment]
I setup the extension to have the log level of verbose, and restarted it a few times. Here are the logs that it generated:

```
2018-02-23 18:14:50.604 -06:00 [Debug/] Finished: Routing Notification "workspace/didChangeConfiguration" in 40ms
2018-02-23 18:14:54.403 -06:00 [Debug/] Finding descriptor for "shutdown"
2018-02-23 18:14:54.406 -06:00 [Debug/] Starting: Routing Request (1) "shutdown"
2018-02-23 18:14:54.406 -06:00 [Debug/] Converting params for Request (1) "shutdown" to "System.Object"
2018-02-23 18:15:01.610 -06:00 [Debug/] Finished: Routing Notification "workspace/didChangeConfiguration" in 35ms
2018-02-23 18:15:04.506 -06:00 [Debug/] Finding descriptor for "shutdown"
2018-02-23 18:15:04.509 -06:00 [Debug/] Starting: Routing Request (1) "shutdown"
2018-02-23 18:15:04.509 -06:00 [Debug/] Converting params for Request (1) "shutdown" to "System.Object"
2018-02-23 18:15:13.085 -06:00 [Debug/] Finished: Routing Notification "workspace/didChangeConfiguration" in 37ms
2018-02-23 18:15:15.960 -06:00 [Debug/] Finding descriptor for "shutdown"
2018-02-23 18:15:15.965 -06:00 [Debug/] Starting: Routing Request (1) "shutdown"
2018-02-23 18:15:15.966 -06:00 [Debug/] Converting params for Request (1) "shutdown" to "System.Object"
```

**Comment by tintoy** [comment]
Ok, thanks - I'll keep digging.

PS. On my Mac, it only ever happens when reloading the window (not when closing and reopening VS Code) so the issue may be partially platform-specific; I'll dig up one of my old Ubuntu VMs to see if I can work out what's going on there...

**Comment by tintoy** [comment]
Thanks either way, and for what it's worth I _have_ managed to reproduce your issue on the Mac (just not reliably yet). I'll keep plugging away at it :)

**Comment by tintoy** [comment]
I _think_ I have a solution (looks like, on Windows, child processes are killed automatically when the VSCode extension process exits but on Mac / Linux they aren't); if you have time, I can walk you through fetching the branch I'm working on and building / running the language server and extension from source to give it a try. Otherwise, I'm happy to build you a VSIX to try out :)

**Comment by alexriedl** [comment]
After cloning the project and switching the the fix/parent-process-exit branch. I ran a `git submodule update --init --recursive`. Then I followed along with the steps to build from [here](https://github.com/tintoy/msbuild-project-tools-vscode/blob/master/docs/BUILDING.md) skipping the debugging steps.

There is no sln file at the root of the project, so the dotnet restore failed. The build-language-server.sh script does it from the lib/server directory, so I did the same thing.

I uninstalled the extension from vscode, and deleted the extension files from ~/.vscode/extensions. Closed vscode, cleaned up any hanging processes, emptied out the log file. Then restarted vscode a few times. Below is the resulting logs:

```
2018-02-24 08:21:09.827 -06:00 [Debug/] Finished: Routing Notification "workspace/didChangeConfiguration" in 38ms
2018-02-24 08:21:09.930 -06:00 [Verbose/] Watching parent process 22275 (language server will automatically shut down if it exits unexpectedly).
2018-02-24 08:21:11.947 -06:00 [Debug/] Finding descriptor for "shutdown"
2018-02-24 08:21:11.950 -06:00 [Debug/] Starting: Routing Request (1) "shutdown"
2018-02-24 08:21:11.950 -06:00 [Debug/] Converting params for Request (1) "shutdown" to "System.Object"
2018-02-24 08:21:11.950 -06:00 [Information/] Server shutdown.
2018-02-24 08:21:17.710 -06:00 [Debug/] Finished: Routing Notification "workspace/didChangeConfiguration" in 37ms
2018-02-24 08:21:17.826 -06:00 [Verbose/] Watching parent process 22559 (language server will automatically shut down if it exits unexpectedly).
2018-02-24 08:21:20.320 -06:00 [Debug/] Finding descriptor for "shutdown"
2018-02-24 08:21:20.322 -06:00 [Debug/] Starting: Routing Request (1) "shutdown"
2018-02-24 08:21:20.322 -06:00 [Debug/] Converting params for Request (1) "shutdown" to "System.Object"
2018-02-24 08:21:20.323 -06:00 [Information/] Server shutdown.
2018-02-24 08:21:26.778 -06:00 [Debug/] Finished: Routing Notification "workspace/didChangeConfiguration" in 38ms
2018-02-24 08:21:26.882 -06:00 [Verbose/] Watching parent process 22847 (language server will automatically shut down if it exits unexpectedly).
2018-02-24 08:21:27.887 -06:00 [Debug/] Finding descriptor for "shutdown"
2018-02-24 08:21:27.889 -06:00 [Debug/] Starting: Routing Request (1) "shutdown"
2018-02-24 08:21:27.889 -06:00 [Debug/] Converting params for Request (1) "shutdown" to "System.Object"
2018-02-24 08:21:27.891 -06:00 [Information/] Server shutdown.
```

Your latest commit to that branch is supposed to log the server version, but I don't see that anywhere in my logs. I'm thinking I missed a step or something.

**Comment by alexriedl** [comment]
I glanced at the server code and noticed it only logs based on the value of environment variables. So I just set those and reran my test. This time I also have logs from the server:

```
2018-02-24 08:59:30.226 -06:00 [Verbose/] Logger initialised.
2018-02-24 08:59:30.300 -06:00 [Verbose/] Language server v"1.0.0" starting in process 26246.
2018-02-24 08:59:30.398 -06:00 [Debug/] Finding descriptor for "initialize"
2018-02-24 08:59:30.622 -06:00 [Debug/] Starting: Routing Request (0) "initialize"
2018-02-24 08:59:30.632 -06:00 [Debug/] Converting params for Request (0) "initialize" to "OmniSharp.Extensions.LanguageServer.Protocol.Models.InitializeParams"
2018-02-24 08:59:30.658 -06:00 [Debug/] Result was "System.Threading.Tasks.Task`1[[OmniSharp.Extensions.LanguageServer.Protocol.Models.InitializeResult, OmniSharp.Extensions.LanguageProtocol, Version=0.7.9.0, Culture=neutral, PublicKeyToken=null]]"
2018-02-24 08:59:30.663 -06:00 [Debug/] Response value was "OmniSharp.Extensions.LanguageServer.Protocol.Models.InitializeResult"
2018-02-24 08:59:30.665 -06:00 [Debug/] Finished: Routing Request (0) "initialize" in 36ms
2018-02-24 08:59:30.699 -06:00 [Debug/] Finding descriptor for "initialized"
2018-02-24 08:59:30.700 -06:00 [Debug/] Finding descriptor for "workspace/didChangeConfiguration"
2018-02-24 08:59:30.714 -06:00 [Debug/] Starting: Routing Notification "initialized"
2018-02-24 08:59:30.715 -06:00 [Debug/] Finished: Routing Notification "initialized" in 1ms
2018-02-24 08:59:30.716 -06:00 [Debug/] Starting: Routing Notification "workspace/didChangeConfiguration"
2018-02-24 08:59:30.716 -06:00 [Debug/] Converting params for Notification "workspace/didChangeConfiguration" to "MSBuildProjectTools.LanguageServer.CustomProtocol.DidChangeConfigurationObjectParams"
2018-02-24 08:59:30.754 -06:00 [Debug/] Finished: Routing Notification "workspace/didChangeConfiguration" in 38ms
2018-02-24 08:59:30.860 -06:00 [Verbose/] Watching parent process 26126 (language server will automatically shut down if it exits unexpectedly).
2018-02-24 08:59:33.426 -06:00 [Debug/] Finding descriptor for "shutdown"
2018-02-24 08:59:33.428 -06:00 [Debug/] Starting: Routing Request (1) "shutdown"
2018-02-24 08:59:33.429 -06:00 [Debug/] Converting params for Request (1) "shutdown" to "System.Object"
2018-02-24 08:59:33.429 -06:00 [Information/] Server shutdown.
2018-02-24 08:59:54.548 -06:00 [Verbose/] Logger initialised.
2018-02-24 08:59:54.622 -06:00 [Verbose/] Language server v"1.0.0" starting in process 26625.
2018-02-24 08:59:54.717 -06:00 [Debug/] Finding descriptor for "initialize"
2018-02-24 08:59:54.921 -06:00 [Debug/] Starting: Routing Request (0) "initialize"
2018-02-24 08:59:54.930 -06:00 [Debug/] Converting params for Request (0) "initialize" to "OmniSharp.Extensions.LanguageServer.Protocol.Models.InitializeParams"
2018-02-24 08:59:54.954 -06:00 [Debug/] Result was "System.Threading.Tasks.Task`1[[OmniSharp.Extensions.LanguageServer.Protocol.Models.InitializeResult, OmniSharp.Extensions.LanguageProtocol, Version=0.7.9.0, Culture=neutral, PublicKeyToken=null]]"
2018-02-24 08:59:54.954 -06:00 [Debug/] Response value was "OmniSharp.Extensions.LanguageServer.Protocol.Models.InitializeResult"
2018-02-24 08:59:54.956 -06:00 [Debug/] Finished: Routing Request (0) "initialize" in 29ms
2018-02-24 08:59:54.987 -06:00 [Debug/] Finding descriptor for "initialized"
2018-02-24 08:59:54.988 -06:00 [Debug/] Finding descriptor for "workspace/didChangeConfiguration"
2018-02-24 08:59:55.001 -06:00 [Debug/] Starting: Routing Notification "initialized"
2018-02-24 08:59:55.002 -06:00 [Debug/] Finished: Routing Notification "initialized" in 1ms
2018-02-24 08:59:55.002 -06:00 [Debug/] Starting: Routing Notification "workspace/didChangeConfiguration"
2018-02-24 08:59:55.003 -06:00 [Debug/] Converting params for Notification "workspace/didChangeConfiguration" to "MSBuildProjectTools.LanguageServer.CustomProtocol.DidChangeConfigurationObjectParams"
2018-02-24 08:59:55.039 -06:00 [Debug/] Finished: Routing Notification "workspace/didChangeConfiguration" in 36ms
2018-02-24 08:59:55.146 -06:00 [Verbose/] Watching parent process 26517 (language server will automatically shut down if it exits unexpectedly).
2018-02-24 09:00:00.731 -06:00 [Debug/] Finding descriptor for "shutdown"
2018-0
2-24 09:00:00.734 -06:00 [Debug/] Starting: Routing Request (1) "shutdown"
2018-02-24 09:00:00.734 -06:00 [Debug/] Converting params for Request (1) "shutdown" to "System.Object"
2018-02-24 09:00:00.734 -06:00 [Information/] Server shutdown.

```

**Comment by tintoy** [comment]
Are you still seeing processes hanging around after restarting with this latest code? It looks to me  like the server claims it is shutting down each time. If so, then perhaps something inside the server is preventing G process termination (e.g. a non-background thread).

The `.sln`file is in `lib/server` BTW.

**Comment by tintoy** [comment]
Ignore that capital G, typing this on my phone where autocorrect is a little over-zealous :)

**Comment by tintoy** [comment]
Those environment variables are set the the client extension, BTW. See `createLanguageClient`in https://github.com/tintoy/msbuild-project-tools-vscode/blob/master/src/extension/extension.ts

**Comment by alexriedl** [comment]
Yes, I am still seeing processes hanging around after restarting.

**Comment by tintoy** [comment]
Ok I'll get a Linux VM set up and try it out. What distro / version are you on?

**Comment by alexriedl** [comment]
I have kubuntu 17.10 installed.

**Comment by tintoy** [comment]
Cheers - I'll get it set up sometime today and see if I can attach a debugger to the stalled processes. I'm the meanwhile, would you mind posting the full output from `ps aux | grep MSBuildProject` and `dotnet --info`?

**Comment by alexriedl** [comment]
```
$ ps aux | grep MSBuildProject
alex      9417  5.5  0.3 3110968 55876 ?       SLl  20:38   0:01 /usr/bin/dotnet /home/alex/.vscode/extensions/tintoy.msbuild-project-tools-0.2.28/out/language-server/MSBuildProjectTools.LanguageServer.Host.dll
alex      9570  0.0  0.0  14352  1076 pts/10   S+   20:38   0:00 grep --color=auto MSBuildProject
```

```
$ dotnet --info
.NET Command Line Tools (2.1.4)

Product Information:
 Version:            2.1.4
 Commit SHA-1 hash:  5e8add2190

Runtime Environment:
 OS Name:     ubuntu
 OS Version:  17.10
 OS Platform: Linux
 RID:         linux-x64
 Base Path:   /usr/share/dotnet/sdk/2.1.4/

Microsoft .NET Core Shared Framework Host

  Version  : 2.0.5
  Build    : 17373eb129b3b05aa18ece963f8795d65ef8ea54
```

**Comment by tintoy** [comment]
Thanks! That'll help :)

**Comment by tintoy** [comment]
Ok have reproduced the issue on a kubuntu 17.10 VM. Will have a look with the debugger and get back to you :)

**Comment by tintoy** [comment]
Ok, @alexriedl, I think I've found the problem (or at least which component is causing the problem). Essentially, server shutdown is blocked [here](https://github.com/OmniSharp/csharp-language-server-protocol/blob/80e1b387ad7ad6d8df7ce944646cb04e07391ea4/src/JsonRpc/ProcessScheduler.cs#L126) (but only on Linux) while waiting for the main dispatcher thread to terminate. I'll chase this up with the OmniSharp LSP folks and let you know what I find.

In the meanwhile, I can produce a build that forcibly terminates the server (using `Environment.Exit()` when the parent process exits, if that would be useful for you.

**Comment by tintoy** [comment]
If you fetch the latest server source (i.e. https://github.com/tintoy/msbuild-project-tools-server/commit/cc107c38d715675537fd93fb83ca3e4d31cb55f8) would you mind trying it again to see if this fixes your problem? It works for me on that Kubuntu VM, so here's hoping :)

If it works, let me know and I'll publish a new release. It's only a temporary workaround but it should unblock you while I sort this out with the LSP folks.

**Comment by tintoy** [comment]
FYI: https://github.com/OmniSharp/csharp-language-server-protocol/pull/78

**Comment by alexriedl** [comment]
With your latest updates processes **DO** get cleaned up when closing and restarting vscode as well as simply reloading the window!

Thanks for focusing on this issue until you got a fix! (even if its only a temporary fix)

**Comment by tintoy** [comment]
Great - I'll put out a new release shortly (v0.2.29)

**Comment by tintoy** [comment]
Hey @alexriedl could you try upgrading to v0.2.29 and see if that works for you?

**Comment by tintoy** [comment]
PS. In the worst case, the server process may take up to 6 seconds to terminate after the window reloads (I wanted to be fairly confident that it's deadlocked, not just running slow, before forcing termination).

**Comment by alexriedl** [comment]
I updated to 0.2.29 and the processes are getting cleaned up

**Comment by tintoy** [comment]
Ok - published to the gallery (the new version may take a couple of minutes to become visible).

**Comment by tintoy** [comment]
I'm going to close this issue for now, but feel free to reopen it if the problem comes up again or you have any questions :)

**Comment by vindicatorr** [comment]
It isn't even just on a restart.
I have 0.2.55 installed and I currently have 8 instances of "~/.vscode-insiders/extensions/tintoy.msbuild-project-tools-0.2.55/out/language-server/MSBuildProjectTools.LanguageServer.Host.dll" left open after vscode is closed.

**Comment by tintoy** [comment]
Hmm, I thought I’d managed to knock this problem on the head (if nothing else, the server now has a watchdog that kills its own process if the client exits without shutting down the server).

Would you mind opening a new issue and adding system info / repro steps so we can troubleshoot?
